criu/tls.c

```diff
@@ -0,0 +1,343 @@
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <linux/limits.h>
+
+#include <gnutls/gnutls.h>
+
+#include "cr_options.h"
+#include "xmalloc.h"
+
+/* Compatability with GnuTLS verson <3.5 */
+#ifndef GNUTLS_E_CERTIFICATE_VERIFICATION_ERROR
+# define GNUTLS_E_CERTIFICATE_VERIFICATION_ERROR GNUTLS_E_CERTIFICATE_ERROR
+#endif
+
+#undef	LOG_PREFIX
+#define LOG_PREFIX "tls: "
+
+#define CRIU_PKI_DIR SYSCONFDIR "/pki"
+#define CRIU_CACERT CRIU_PKI_DIR "/CA/cacert.pem"
+#define CRIU_CACRL CRIU_PKI_DIR "/CA/cacrl.pem"
+#define CRIU_CERT CRIU_PKI_DIR "/criu/cert.pem"
+#define CRIU_KEY CRIU_PKI_DIR "/criu/private/key.pem"
+
+#define SPLICE_BUF_SZ_MAX (PIPE_BUF * 100)
+
+#define tls_perror(msg, ret) pr_err("%s: %s\n", msg, gnutls_strerror(ret))
+
+static gnutls_session_t session;
+static gnutls_certificate_credentials_t x509_cred;
+static int tls_sk = -1;
+static int tls_sk_flags = 0;
+
+void tls_terminate_session()
+{
+	int ret;
+
+	if (!opts.tls)
+		return;
+
+	if (session) {
+		do {
+			/* don't wait for peer to close connection */
+			ret = gnutls_bye(session, GNUTLS_SHUT_WR);
+		}
+		while(ret == GNUTLS_E_AGAIN || ret == GNUTLS_E_INTERRUPTED);
+		gnutls_deinit(session);
+	}
+
+	tls_sk = -1;
+	if (x509_cred)
+		gnutls_certificate_free_credentials(x509_cred);
+}
+
+ssize_t tls_send(const void *buf, size_t len, int flags)
+{
+	int ret;
+
+	tls_sk_flags = flags;
+	do {
+		ret = gnutls_record_send(session, buf, len);
+	}
+	while (ret == GNUTLS_E_INTERRUPTED || ret == GNUTLS_E_AGAIN);
+	tls_sk_flags = 0;
+
+	if (gnutls_error_is_fatal(ret)) {
+		tls_perror("Failed to send data", ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+/*
+ * Read data from a file descriptor, then encrypt and send it with GnuTLS.
+ * This function is used for cases when we would otherwise use splice()
+ * to transfer data from PIPE to TCP socket.
+ */
+int tls_send_data_from_fd(int fd, unsigned long len)
+{
+	ssize_t copied;
+	unsigned long buf_size = min(len, (unsigned long)SPLICE_BUF_SZ_MAX);
+	void *buf = xmalloc(buf_size);
+
+	if (!buf)
+		return -1;
+
+	while (len > 0) {
+		copied = read(fd, buf, buf_size);
+		if (copied < 0) {
+			pr_perror("Can't read from pipe");
+			goto err;
+		}
+
+		if (copied == 0)
+			break;
+
+		int ret, sent;
+		for(sent = 0; sent < copied; sent += ret) {
+			ret = tls_send((buf + sent), (copied - sent), 0);
+			if (ret < 0) {
+				tls_perror("Failed sending data", ret);
+				goto err;
+			}
+		}
+		len -= copied;
+	}
+err:
+	xfree(buf);
+	return (len > 0);
+}
+
+ssize_t tls_recv(void *buf, size_t len, int flags)
+{
+	int ret;
+
+	tls_sk_flags = flags;
+	do {
+		ret = gnutls_record_recv(session, buf, len);
+	}
+	while (ret == GNUTLS_E_INTERRUPTED || ret == GNUTLS_E_AGAIN);
+	tls_sk_flags = 0;
+
+	if (ret == GNUTLS_E_REHANDSHAKE) {
+		tls_perror("Re-handshake is not supported", ret);
+		return -1;
+	}
+
+	if (gnutls_error_is_fatal(ret)) {
```

Comment:
do we need to report non-fatal errors?

Reply:
[non-fatal errors](https://github.com/gnutls/gnutls/blob/0a47c08db07d0a01018aec34d10a294aed79fdd1/lib/errors.c#L442L461) are:
```C
static const gnutls_error_entry non_fatal_error_entries[] = {
	ERROR_ENTRY(N_("Success."), GNUTLS_E_SUCCESS),
	ERROR_ENTRY(N_("A TLS warning alert has been received."),
		    GNUTLS_E_WARNING_ALERT_RECEIVED),
	ERROR_ENTRY(N_("A heartbeat pong message was received."),
		    GNUTLS_E_HEARTBEAT_PONG_RECEIVED),
	ERROR_ENTRY(N_("A heartbeat ping message was received."),
		    GNUTLS_E_HEARTBEAT_PING_RECEIVED),
	ERROR_ENTRY(N_("Resource temporarily unavailable, try again."),
		    GNUTLS_E_AGAIN),
	ERROR_ENTRY(N_("The transmitted packet is too large (EMSGSIZE)."),
		    GNUTLS_E_LARGE_PACKET),
	ERROR_ENTRY(N_("Function was interrupted."), GNUTLS_E_INTERRUPTED),
	ERROR_ENTRY(N_("Rehandshake was requested by the peer."),
		    GNUTLS_E_REHANDSHAKE),
	ERROR_ENTRY(N_("Re-authentication was requested by the peer."),
		    GNUTLS_E_REAUTH_REQUEST),
	/* Only non fatal (for handshake) errors here */
	{NULL, NULL, 0}
};
```
Currently re-authentication, re-handshake and alerts are not used/expected, so I don't think that logging non-fatal errors would be very useful.

Reply:
Since non-fatal errors are not expected I changed this condition for tls_send() and tls_recv() to return -1 in case of any error.